src/js/trove/matrix.js

```diff
@@ -0,0 +1,1299 @@
+            /*For Gram Schmidt we operate on columns as vectors so we just create an array of vectors from matrix */
+            var ans_vectors = new Array(self.$w);
+            function colToVector(cn) {
+                return makeVector(new Array(self.$h).fill(0).map((v, i) => get1dElem(self, i, cn)));
+            }
+            ans_vectors[0] = colToVector(0);
+            var mags = new Array(self.$w);
+            function magSqr(n) {
+                tmp = ans_vectors[n].$underlyingMat.reduce((a, v) => runtime.plus(a, runtime.times(v, v)), 0);
+                mags[n] = tmp;
+            }
+            magSqr(0);
+
+            for (col = 1; col < self.$w; col++) {
+                var currVector = colToVector(col);
+
+                for (prev = 0; prev < col; prev++) {
+
+                    sclr = runtime.divide(dotVector(currVector, ans_vectors[prev]), mags[prev]);
+                    currVector = minusVector(currVector, scaleVector(ans_vectors[prev], sclr));
+                }
+                ans_vectors[col] = currVector
+                magSqr(col);
+            }
+            mags = mags.map(v => runtime.num_sqrt(v));
+            var new_arr = new Array(self.$h * self.$w);
+            for (i = 0; i < self.$h; i++) {
+                for (j = 0; j < self.$w; j++) {
+                    new_arr[get1dpos(i, j, self.$w)] = runtime.divide(ans_vectors[j].$underlyingMat[i], mags[j]);
+                }
+            }
+
+            return makeMatrix(self.$h, self.$w, new_arr);
+
+
+
+        }
+        //Finds QR decomposition using Gram Schmidt
+        var qrMatrix = function (self) {
+            arity(1, arguments, "qr-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "qr-mat", self, annMatrix);
+            q = Gram_Schmidt(self);
+            r = multMatrix(transposeMatrix(q), self);
+            return runtime.makeTuple([q, r]);
+        }
+        //Finds pth norm of matrix
+        var normMatrix = function (self, p) {
+            arity(2, arguments, "norm-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "norm-mat", self, annMatrix, p, runtime.Number);
+            sum = self.$underlyingMat.reduce((a, v) => runtime.plus(a, runtime.num_expt(v, p)), 0);
+            return runtime.num_expt(sum, runtime.divide(1, p));
+        }
+        //Creates matrix of size (h,w) by applying f
+        var buildMatrix = function (h, w, f) {
+            arity(3, arguments, "build-mat", false);
+            runtime.checkArgsInternalInline("Matrix", "build-mat", h, runtime.Number, w, runtime.Number, f, runtime.Function);
+            posInteger(h, w);
+            return runtime.safeCall(
+                function () {
+                    return runtime.raw_array_build(F(function (idx) {
+                        res = f.app((idx - (idx % w)) / w, idx % w);
+                        if (!jsnum.isPyretNumber(res)) {
```

Comment:
Not quite -- you need this test of the result to be in the continuation of `f.app`: basically, you have to assume that `f` might bail out at any point in its execution, and `safeCall` says, "ok, Pyret, look -- when you _do_ finally finish running `f`, you need to follow up by doing _this_."  So you need to say `safeCall(f.app(...), function(res) { if (!isNumber) { ... } else { return res } })`